onnxruntime/core/util/math_cpu.cc

```diff
@@ -269,32 +269,51 @@ SPECIALIZED_SET(uint16_t);
 #undef SPECIALIZED_SET
 
 template <typename T>
-static void Im2colWithEqualPadding(int64_t output_h, int64_t output_w, const T* data_im, int64_t channels,
-                                   int64_t height, int64_t width, int64_t kernel_h, int64_t kernel_w,
-                                   int64_t dilation_h, int64_t dilation_w, int64_t pad_t, int64_t pad_l,
-                                   int64_t stride_h, int64_t stride_w, T* data_col, T padding_value) {
+void Im2col<T, StorageOrder::NCHW>::operator()(const T* data_im, int64_t channels, int64_t height,
+                                               int64_t width, int64_t kernel_h, int64_t kernel_w,
+                                               int64_t dilation_h, int64_t dilation_w, int64_t pad_t,
+                                               int64_t pad_l, int64_t pad_b, int64_t pad_r, int64_t stride_h,
+                                               int64_t stride_w, T* data_col, T padding_value) {
+  const int64_t output_h = (height + pad_b + pad_t - (dilation_h * (kernel_h - 1) + 1)) / stride_h + 1;
+  const int64_t output_w = (width + pad_l + pad_r - (dilation_w * (kernel_w - 1) + 1)) / stride_w + 1;
+
   // From Intel, https://github.com/BVLC/caffe/pull/3536
-  int64_t pad_h = pad_t;
-  int64_t pad_w = pad_l;
   int64_t channel_size = height * width;
   for (int64_t channel = channels; channel--; data_im += channel_size) {
     for (int64_t kernel_row = 0; kernel_row < kernel_h; kernel_row++) {
       for (int64_t kernel_col = 0; kernel_col < kernel_w; kernel_col++) {
-        int64_t input_row = -pad_h + kernel_row * dilation_h;
```

Comment:
Can it get better data locality if iterating output_h and ouput_w firstly?

Reply:
The access pattern here (and in the old version) is to stream out the im2col buffer in sequential order. Changing the loop order breaks that and may get better read locality at the cost of this write locality. There very well could be better patterns than what is here, but this change is smaller/faster than what existed before.

Anything I need to change before getting this in?